hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/TestSCMContainerPlacementRackAware.java

```diff
@@ -524,4 +527,28 @@ public void testvalidateContainerPlacementSingleRackCluster() {
     assertTrue(stat.isPolicySatisfied());
     assertEquals(0, stat.misReplicationCount());
   }
+
+  @Test
+  public void testOutOfServiceNodesNotSelected() {
+    // Set all the nodes to out of service
+    for (DatanodeInfo dn : dnInfos) {
+      dn.setNodeStatus(new NodeStatus(DECOMMISSIONED, HEALTHY));
+    }
+
+    for (int i=0; i<10; i++) {
+      // Set a random DN to in_service and ensure it is always picked
+      int index = new Random().nextInt(dnInfos.size());
+      dnInfos.get(index).setNodeStatus(NodeStatus.inServiceHealthy());
+      try {
+        List<DatanodeDetails> datanodeDetails =
+            policy.chooseDatanodes(null, null, 1, 0, 0);
+        Assert.assertEquals(dnInfos.get(index), datanodeDetails.get(0));
+      } catch (SCMException e) {
+        // If we get SCMException: No satisfied datanode to meet the ... this is
```

Comment:
Should we add a check at the end to make sure that `policy.chooseDatanodes` worked at least once?

Reply:
I think that would risk making the test flaky. When there are a large number of datanodes (say 15) and we put all out of service but 1, there is a reasonably high chance it will never find a node to use in 10 tries. I think it only retries 3 times inside the policy so it has a 1 in 15 chance on the first try, and 1 in 14 on the second try and 1 in 13 on the 3rd. Its actually more likely to fail than find a node.

I could argue that is a problem with the provider, but that logic was put in to avoid infinite loops, and we are not trying to fix the provider here. Probably the provider should keep trying until all nodes have been tried, but that would be a topic for another Jira I think.

Reply:
Should we then change the setup and retry count to make sure this test actually tests as intended. A successful run here would not imply that the only node selected is the one that is healthy.

Reply:
The retry count in the provider is hard coded at the moment, and this is not a realistic real-world scenario. If there are only a few healthy nodes out of 100's in a cluster, then the cluster has much bigger problems. If we want to change the behaviour of the provider, then we should discuss that in a new Jira.

The contact of the provider is to provide the number of requested nodes, or throw the SCM Exception. So either we get the single IN_SERVICE node, or we get the exception. The key is that we never get the out_of_service node, which is the defect the original change was fixing. If you removed the fix for the original issue, this test would always fail as it is most likely to find the out_of_service node each time, and over 10 tries in the loop it would almost certainly find an out_of_service node and fail.

I cannot think of a better way to test this without changing the provider code (which I don't want to do in this Jira), but if you have a better idea that will never be flaky I am happy to use it.

Reply:
The networkTopology doesn't have the node healthy information currently.  There is chance that chooseDatanodes will throw exception after 3 internal retries that it cannot find a proper node that is healthy if there are two many unhealthy nodes in the cluster.  One improvement I can think of is add the the healthy status and storage space full status  into the NetworkTopology.  And I agree with @sodonnel ,  we can discuss it in another JIRA.